File: taurus-protect-sdk-typescript/src/models/webhook.ts
/**
 * Webhook models for Taurus-PROTECT SDK.
 */

import type { CursorPage, CursorPageOptions } from './pagination';

/**
 * Webhook status enum.
 */
export enum WebhookStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  FAILED = 'FAILED',
}

/**
 * Webhook information.
 */
export interface Webhook {
  /** Unique webhook identifier */
  readonly id?: string;
  /** Webhook URL */
  readonly url?: string;
  /** Event types subscribed to */
  readonly events?: string[];
  /** Webhook status */
  readonly status?: WebhookStatus;
  /** Secret for HMAC signature verification */
  readonly secret?: string;
  /** Creation date */
  readonly createdAt?: Date;
  /** Last modification date */
  readonly updatedAt?: Date;
  /** Number of consecutive failures */
  readonly failureCount?: number;
  /** Last failure message */
  readonly lastFailureMessage?: string;
}

/**
 * Options for listing webhooks. A cursor list: `pageSize` 1-100 (default 20) and the
 * `cursor` of a previous page.
 */
export interface ListWebhooksOptions extends CursorPageOptions {
  /** Filter by webhook type */
  type?: string;
  /** Filter by URL */
  url?: string;
  /** Sort order (ASC or DESC) */
  sortOrder?: 'ASC' | 'DESC';
}

/**
 * A page of webhooks.
 */
export interface ListWebhooksResult {
  /** The webhooks of this page */
  readonly items: Webhook[];
  /** Cursor pagination: continue with `cursor: pagination.nextCursor` while `hasMore` */
  readonly pagination: CursorPage;
}


/**
 * Request for creating a webhook.
 */
export interface CreateWebhookRequest {
  /** Webhook URL */
  url: string;
  /** Event types to subscribe to */
  events: string[];
}

File: taurus-protect-sdk-go/pkg/protect/model/webhook.go
package model

import "time"

// Webhook represents a webhook configuration in Taurus-PROTECT.
type Webhook struct {
	// ID is the unique identifier for the webhook.
	ID string `json:"id"`
	// Type is the webhook type (e.g., "request", "transaction").
	Type string `json:"type"`
	// URL is the endpoint URL where webhook events are sent.
	URL string `json:"url"`
	// Status is the webhook status (e.g., "enabled", "disabled").
	Status string `json:"status"`
	// Secret is the webhook secret (only returned on creation).
	Secret string `json:"secret,omitempty"`
	// TimeoutUntil is when the webhook timeout expires (if timed out).
	TimeoutUntil time.Time `json:"timeout_until,omitempty"`
	// CreatedAt is when the webhook was created.
	CreatedAt time.Time `json:"created_at"`
	// UpdatedAt is when the webhook was last updated.
	UpdatedAt time.Time `json:"updated_at"`
}

// CreateWebhookRequest contains the parameters for creating a webhook.
type CreateWebhookRequest struct {
	// Type is the webhook type (e.g., "request", "transaction").
	Type string `json:"type"`
	// URL is the endpoint URL where webhook events will be sent.
	URL string `json:"url"`
}

// CreateWebhookResult contains the result of creating a webhook.
type CreateWebhookResult struct {
	// Webhook is the created webhook.
	Webhook *Webhook `json:"webhook"`
	// Secret is the webhook secret (only provided once on creation).
	Secret string `json:"secret"`
}

// ListWebhooksOptions contains options for listing webhooks.
type ListWebhooksOptions struct {
	// Type filters webhooks by type.
	Type string
	// URL filters webhooks by URL.
	URL string
	// PageSize is the page size: 0 selects DefaultPageSize, above MaxPageSize is an error.
	PageSize int64
	// Cursor is a previous page's Page.NextCursor; the SDK then requests the NEXT page.
	Cursor string
	// CurrentPage and PageRequest (FIRST, PREVIOUS, NEXT, LAST) page by hand; neither can be
	// combined with Cursor.
	CurrentPage string
	PageRequest string
	// SortOrder specifies the sort order (ASC or DESC). Default is DESC.
	SortOrder string
}

// ListWebhooksResult contains the result of listing webhooks.
type ListWebhooksResult struct {
	// Webhooks is the list of webhooks.
	Webhooks []*Webhook `json:"webhooks"`
	// Page continues the list: pass Page.NextCursor as the next Cursor until HasMore is false.
	Page CursorPage `json:"page"`
}
